pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+            TotalRoots::<T>::put(id + 1.into());
+            Self::deposit_event(RawEvent::RootCreated(id, who));
+            Ok(())
+        }
+        
+        /// Form index and store input to redzones, creates area struct if it doesnt exist
+        #[weight = <T as Trait>::WeightInfo::zone_add()]
+        pub fn zone_add(origin, 
+                        rect: Rect2D<Point2D<T::Coord>>,
+                        height: u16,
+                        root_id: T::RootId) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            ensure!(RootBoxes::<T>::contains_key(root_id), Error::<T>::RootDoesNotExist);
+            ensure!(height > 1, Error::<T>::InvalidData);
+            // TODO calc required area in root from rect, rn no overlap checks
+            let area_id = Self::detect_touch(RootBoxes::<T>::get(root_id), rect.north_west);
+                        
+            let area = if AreaData::<T>::contains_key(root_id, area_id) {
+                AreaData::<T>::get(root_id, area_id)
+            } else {
+                AreaData::<T>::insert(root_id, area_id, Area::new(GREEN_AREA, 0));
+                Area::new(GREEN_AREA, 0)
+            };
+
+            ensure!(area.area_type == GREEN_AREA, Error::<T>::ForbiddenArea); 
+            let id = Self::form_index(root_id, area_id, area.child_amount); 
+            let zone = ZoneOf::<T>::new(id, rect, height);
+            RedZones::<T>::insert(id, zone);
+            AreaData::<T>::mutate(root_id, area_id, |ar| {
+                ar.child_amount += 1;
+            });
+            Self::deposit_event(RawEvent::ZoneCreated(root_id, area_id, id, who));
+            Ok(())
+        }
+
+        /// Changes area type
+        #[weight = <T as Trait>::WeightInfo::change_area_type()]
+        pub fn change_area_type(origin, 
+                                root_id: T::RootId, 
+                                area_id: T::AreaId, 
+                                area_type: u8) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            ensure!(AreaData::<T>::contains_key(root_id, area_id), Error::<T>::NotExists);
+            AreaData::<T>::mutate(root_id, area_id, |ar| {
+                ar.area_type = area_type;
+            });
+            Self::deposit_event(RawEvent::AreaTypeChanged(area_type, area_id, root_id, who));
+            Ok(())
+        }
+    }
+}
+
+// Module allows  use  common functionality by dispatchables
+impl<T: Trait> Module<T> {
+    // Implement module function.
+    // Public functions can be called from other runtime modules.
+    // TODO implement safe math
+    fn detect_touch(root_box: RootBoxOf<T>, touch: Point2D<T::Coord>) -> T::AreaId {
```

Comment:
What does this fn do? What is "touch"? Probably needs comments or/and better naming.